File: backend/Models/WorkspaceMembership.cs
namespace backend.Models;

public enum WorkspaceRole
{
    Owner,
    Admin,
    Member,
}

public class WorkspaceMembership
{
    public int Id { get; set; }
    public int UserId { get; set; }
    public int WorkspaceId { get; set; }
    public WorkspaceRole Role { get; set; } = WorkspaceRole.Member;
    public DateTime? JoinedAt { get; set; } = null;
    public DateTime? InvitedAt { get; set; } = null;

    // Navigation properties
    public User User { get; set; } = null!;
    public Workspace Workspace { get; set; } = null!;
}

File: frontend/src/shared/types/workspacemembership.ts
import { type UserReadDto } from './user';
import type { WorkspaceReadDto } from './workspace';

interface WorkspaceMembershipBase {
    userId: number;
    workspaceId: number;
}
export interface WorkspaceMembershipInviteDto {
  email: string;
  workspaceId: number;
  role: WorkspaceRole;
}

export interface WorkspaceMembershipInvitationReadDto {
  id: number;
  role: WorkspaceRole;
  invitedAt: string;
  workspace: WorkspaceReadDto;
}

export interface WorkspaceMembershipReadDto extends WorkspaceMembershipBase {
  role: WorkspaceRole;
  joinedAt: string | null;
  user: UserReadDto;
}

export interface WorkspaceMembershipCreateDto extends WorkspaceMembershipBase {
  role: WorkspaceRole;
}

export interface WorkspaceMembershipUpdateDto {
  role: WorkspaceRole;
}

export const WorkspaceRole = {
    Owner: 'Owner',
    Admin: 'Admin',
    Member: 'Member'
} as const;

export type WorkspaceRole = typeof WorkspaceRole[keyof typeof WorkspaceRole];
